### PR TITLE
Fea: Detective hat detach

### PR DESCRIPTION
**Pull request by mzegar**

### CODE REVIEW

File: code/Player/Player.Clothes.cs [hunk 1]
```diff
@@ -35,16 +35,19 @@ public void AttachClothing( string path )
...
-				clothingPiece.Delete();
-			}
+			if ( clothingPiece.ModelPath != path )
+				continue;
+
+			clothingPiece.Delete();
+			return true;
 		}
+
+		return false;
```
Comment (mzegar) [inline code/Player/Player.Clothes.cs:50]: Might be able to change this if you had any ideas

File: code/Other/DetectiveHat.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+using Sandbox;
+
+namespace TTT;
+
+[HideInEditor]
+[EditorModel( "models/detective_hat/detective_hat.vmdl" )]
+[Title( "Detective Hat" )]
+public partial class DetectiveHat : Prop, IEntityHint
```
Comment (kolexxx) [inline code/Other/DetectiveHat.cs:8]: Should derive from `BaseClothing`.

**Comment by kolexxx** [comment]
My things that I plan on changing. `DetectiveHat` isn't needed, we should only be using the `Clothing` class. Make hands work with every clothing item, not only the detective hat. Also apply some force to the hat.

**Comment by mzegar** [comment]
It won't look good with the other clothing items. Their 3d models will look awful off the player.

**Comment by kolexxx** [comment]
> It won't look good with the other clothing items. Their 3d models will look awful off the player.

We will only be detaching the detective hat for now. My whole point was to make the system generic and not just the detective hat.

**Comment by mzegar** [comment]
Sure, as long as you can still spawn the hat as a hammer entity, go wild

**Comment by kolexxx** [comment]
> Sure, as long as you can still spawn the hat as a hammer entity, go wild

Mappers shouldn't be placing detective hats around the map.

**Comment by mzegar** [comment]
If they can place cigars around the map, let them put a detective hat

File: code/Player/Corpse.cs [hunk 3]
```diff
@@ -73,7 +73,18 @@ public Corpse( Player player )
...
 		foreach ( var clothing in Player.Clothes.ToArray() )
-			clothing.SetParent( this, true );
+		{
+			if ( clothing.Model == Detective.Hat && (HitboxGroup)GetHitboxGroup( Player.LastDamage.HitboxIndex ) == HitboxGroup.Head )
+			{
+				clothing.Parent = null;
+				clothing.EnableAllCollisions = true;
+				clothing.PhysicsEnabled = true;
+				clothing.UsePhysicsCollision = true;
+				clothing.PhysicsGroup.AddVelocity( Player.LastDamage.Force.Length *  Vector3.Up );
```
Comment (mzegar) [inline code/Player/Corpse.cs:83]: Maybe we should just wrap this up into a public method inside of `class Clothing`

File: code/Items/Equipment/Hands/Hands.GrabbableProp.cs [hunk 4]
```diff
@@ -6,10 +6,10 @@ namespace TTT;
 public class GrabbableProp : IGrabbable
 {
 	private ModelEntity GrabbedEntity { get; set; }
-	public string PrimaryAttackHint => GrabbedEntity.IsValid() ? "Throw" : string.Empty;
-	public string SecondaryAttackHint => GrabbedEntity.IsValid() ? "Drop" : string.Empty;
+	public virtual string PrimaryAttackHint => GrabbedEntity.IsValid() ? "Throw" : string.Empty;
+	public virtual string SecondaryAttackHint => GrabbedEntity.IsValid() ? "Drop" : string.Empty;
```
Comment (mzegar) [inline code/Items/Equipment/Hands/Hands.GrabbableProp.cs:10]: All the virtuals in this class can be removed since we got rid of `DetectiveHat.cs`

**Comment by mzegar** [comment]
🤯

**Pull request merged**